community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/pipes/ExpandIntoPipe.scala

```diff
@@ -35,33 +43,31 @@ case class ExpandIntoPipe(source: Pipe,
                           types: LazyTypes)(val estimatedCardinality: Option[Double] = None)
                          (implicit pipeMonitor: PipeMonitor)
   extends PipeWithSource(source, pipeMonitor) with RonjaPipe {
-
   self =>
 
+  private final val CACHE_SIZE = 100000
+
   protected def internalCreateResults(input: Iterator[ExecutionContext], state: QueryState): Iterator[ExecutionContext] = {
+    //cache of known connected nodes
+    val relCache = new RelationshipsCache(CACHE_SIZE)
+
     input.flatMap {
       row =>
         val fromNode = getRowNode(row, fromName)
         fromNode match {
           case fromNode: Node =>
-            val relationships: Iterator[Relationship] = state.query.getRelationshipsForIds(fromNode, dir, types.types(state.query))
-            val toNode: Node = getRowNode(row, toName)
-
-            if (toNode == null)
-              Iterator.empty
-            else
-              new PrefetchingIterator[ExecutionContext] {
-                override def fetchNextOrNull(): ExecutionContext = {
-                  while (relationships.hasNext) {
-                    val rel = relationships.next()
-                    val other = rel.getOtherNode(fromNode)
-                    if (toNode == other) {
-                      return row.newWith2(relName, rel, toName, toNode)
-                    }
-                  }
-                  null
-                }
-              }.asScala
+            val toNode = getRowNode(row, toName)
+
+            if (toNode == null) Iterator.empty
+            else {
+              if (relCache.notConnected(fromNode, toNode)) Iterator.empty
+              else {
+                //Check cache for connections otherwise, iterate through the relationships
+                relCache.get(fromNode, toNode)
```

Comment:
Only do the cache lookup once, work with the returned seq to check for empty == not connected